server/context.ts

```diff
@@ -91,14 +100,54 @@ async function getAndUpdateSession(
   if (Object.keys(sessionUpdates).length > 0) {
     console.log("Updating session:", sessionUpdates);
 
-    prisma.session
-      .update({
-        where: { id: sessionId },
-        data: sessionUpdates,
-      })
-      .catch((error) => {
-        console.error("Error updating session:", error);
+    // Use the secure database function with SECURITY DEFINER
+    try {
+      // PgBouncer in transaction mode requires special handling
+      const authPrisma = new PrismaClient({
+        datasources: {
+          db: {
+            url: process.env.POSTGRES_PRISMA_URL,
+          },
+        },
       });
+      
+      // Maximum retry attempts
+      const maxRetries = 3;
+      let attempts = 0;
+      let success = false;
+      
+      while (attempts < maxRetries && !success) {
+        try {
+          // Call the upsert_session function using a simpler query format
+          // This works better with PgBouncer
+          await authPrisma.$executeRawUnsafe(`
```

Comment:
As we call this function from the `createContext()` above, which has an authenticated PrismaClient instance, is there any reason to use `$executeRawUnsafe` instead of calling `prisma.session.update`?

Reply:
this has been refactored